pkg/controller/master/backup/backup.go

```diff
@@ -226,6 +247,65 @@ func (h *Handler) getBackupSource(vmBackup *harvesterv1.VirtualMachineBackup) (*
 	return sourceVM, nil
 }
 
+func (h *Handler) getBackupSourceInstance(vmBackup *harvesterv1.VirtualMachineBackup) (*kubevirtv1.VirtualMachineInstance, error) {
+	if vmBackup.Spec.Source.Kind != kubevirtv1.VirtualMachineGroupVersionKind.Kind {
+		return nil, fmt.Errorf("unsupported source: %+v", vmBackup.Spec.Source)
+	}
+
+	sourceVMI, err := h.vmisCache.Get(vmBackup.Namespace, vmBackup.Spec.Source.Name)
+	if err != nil {
+		return nil, err
+	}
+	if sourceVMI.DeletionTimestamp != nil {
+		return nil, fmt.Errorf("vm %s/%s is being deleted", vmBackup.Namespace, vmBackup.Spec.Source.Name)
+	}
+
+	return sourceVMI, nil
+}
+
+func (h *Handler) tryFreezeFS(backup *harvesterv1.VirtualMachineBackup) {
```

Comment:
why we are not returning the error here? it should retry the freeze before creating the volumeSnapshot object.

Reply:
I initially didn't add a retry so that freezing the FS is best-effort but doesn't block the snapshot from being carried out.

Should it retry only a certain amount of times or indefinitely? What kind of behavior would you prefer?

Reply:
I believe the fs freeze should be a required operation so that the snapshotContent will always be expected and more consistent. 
The controller framework will reconcile it as long as you have returned an appropriate error, I would suggest we add the error message to the vmBackup status and clean it up after it was succeeded.

Reply:
For this specific scenario, it is better not to return error and retry, a couple of considerations:
(1) the freeze/thaw happens before the snapshot/backup, if VM is keeping writing, this action may not help too much, but at least, no harm
(2) it is difficult to define an unified timeout param, in certain environment, if fail, it may always fail, e.g., timeout is not enough, or low level fs / driver return error due to other reasons, the retry will block this feature

(3) besides this independent enhancement, we need to further investigate & enhance the end to end backup/snapshot

![image](https://github.com/harvester/harvester/assets/31133476/9650eeb5-73ad-40ad-b30b-7f4cc7d6b29e)

we may log a warning / k8s event when freeze fails

Reply:
@w13915984028 @connorkuehl Adding a proper log and event would be helpful, but for each API call or controller reconciliation, we need to make sure the business logic and result should be as idempotent as possible.
The fsfreeze will help to halt new access to the filesystem and creates a stable image on disk, so at least it will help to make sure the snapshot content is more expected than the previous random behavior, and in regards to the above concerns:
- since this is a live snapshot/backup, so the content taken by the snapshot would only be the best effort; if users want  a consistent one, shutdown the VM and taking either a snapshot or backup probably is the best option(and we can document this)。
- we need to add proper status if the fsfreeze call fails, so the user can decide whether to delete and create a new one, or wait for the controller to do the reconciliation(instead of skipping the error as the user may never check the log/event).
- TODO: you probably don't want to call fs to freeze when the VM is shut down

Reply:
> TODO: you probably don't want to call fs to freeze when the VM is shut down

I think this is currently the case, since the function returns early if there's no VMI.

Reply:
> but for each API call or controller reconciliation, we need to make sure the business logic and result should be as idempotent as possible.

Just to make sure I'm parsing this correctly, does this mean that if the backup controller confirms the qemu-guest-agent is present and it tries to freeze the FS but it errors, then subsequent reconciliation attempts should remember that the first attempt was made with fsfreeze and so it should always try to freeze the FS even if the qemu-guest-agent is no longer running?

So instead of this:

```mermaid
flowchart TD
    backupchange(OnBackupChange)
    createvolsnap(h.createVolumeSnapshot)
    hasga{Has guest agent?}
    putfrz(PUT freeze subresource)
    isrun{Is running/has VMI}
    
    Start-->backupchange
    backupchange --> reconcileVolumeSnapshots
    reconcileVolumeSnapshots --> isrun
    isrun-- yes -->hasga
    isrun-- no --> createvolsnap
    hasga-- yes --> putfrz
    hasga-- no --> createvolsnap
    putfrz-- fail --> Enqueue
    putfrz-- success --> createvolsnap
    Enqueue-.->backupchange
```

It should behave like this:

```mermaid
flowchart TD
    backupchange(OnBackupChange)
    createvolsnap(h.createVolumeSnapshot)
    recvolsnap(reconcileVolumeSnapshots)
    storega(persist fsfreeze strategy=true somewhere?)
    storenoga(persist fsfreeze strategy=false somewhere?)
    storenoga2(persist fsfreeze strategy=false somewhere?)
    hasga{Has guest agent?}
    putfrz(PUT freeze subresource)
    hasfrzstrat{Has fsfreeze strategy?}
    fetchfrzstrat(fetch persisted freeze strategy)
    isrun{Is running/has VMI}
    
    Start-->backupchange
    backupchange --> recvolsnap
    recvolsnap --> hasfrzstrat
    hasfrzstrat-- no -->isrun
    isrun-- yes -->hasga
    isrun-- no -->storenoga2
    storenoga2-->createvolsnap
    hasfrzstrat-- yes -->fetchfrzstrat
    fetchfrzstrat-- fsfreeze=true -->putfrz
    fetchfrzstrat-- fsfreeze=false -->createvolsnap
    putfrz-. error .-> Enqueue
    putfrz-- ok -->createvolsnap
    Enqueue-.->backupchange
    
    hasga-- no -->storenoga
    hasga-- yes -->storega
    storenoga-->fetchfrzstrat
    storega-->fetchfrzstrat
```

Reply:
@connorkuehl the new flow looks correct to me, and I think we could just store it in the VMBackup annotations.

Reply:
Thanks, @guangbochen! I've pushed a revision that implements the new flow.

To my other reviewers, FYI: the code has changed in a nontrivial but not hugely substantial way since the revision you approved. (in case this changes your approval) CC: @w13915984028 @WebberHuang1118 @ibrokethecloud @markhillgit